test/README.md

```diff
@@ -0,0 +1,125 @@
+organised very poorly (inconsistent naming, inconsistent file structure,
+inconsistent use of testing directives, etc.).
+
+This document aims to clarify how we conduct testing in the Perpetual Pools
+system and act as a procedural guide for how to write additional tests when
+contributing to the codebase. Specifically, this document aims to address the
+following questions:
+
+ - How should tests be laid out? File-per-contract, etc.
+ - What testing frameworks do we use?
+ - How should the tests themselves look?
+
+## Test Suite Structure ##
+
+### All Tests Under `test` ###
+
+All tests -- regardless of their type (e.g., unit, integration, etc.) -- must
+reside under the `test` directory.
+
+### One-File-One-Contract ###
+
+All unit tests should be arranged such that all tests for a given module are in
+the same file. The most obvious choice for what constitutes a "module" is a
+smart contract.
+
+For example, if the (fictional) contract `PermissionedDeployer` is being added
+to the codebase, its unit tests should reside in a file called
+`PermissionedDeployer.spec.ts` under the `test` directory.
+
+## Test Layout ##
+
+### Use BDD Testing Nomenclature ###
+
+Perpetual Pools uses [Mocha](https://mochajs.org) as its testing framework.
+Broadly speaking, Mocha exposes three different primitives for writing tests:
+
+ - `describe`
+    - Used to *describe* a given component of the system being tested
+    - Can be nested
+ - `context`
+    - Used to describe the situation being considered
+ - `it`
+    - Used to make an assertion about the behaviour of the system
+
+Additionally, the ability to nest these primitives must be used eagerly. For
+example,
+
+```typescript
+describe("PermissionedDeployer", async () => {
+    describe("get", async () => {
+        context("When called with an offset less than zero", async () => {
+            it("Reverts", async () => {
+                let negativeOffset: BigNumberish = -1;
+                await expect(permissionedDeployer.get(negativeOffset)).to.be.revertedWith("PD: Offset must be non-negative")
+            })
+        })
+
+        context("When called with an offset equal to zero", async () => {
+            it("Reverts", async () => {
+                let zeroOffset: BigNumberish = 0
+                let expectedResult: BigNumberish = 100 /* arbitrary */
+
+                expect(await permissionedDeployer.get(zeroOffset)).to.eq(expectedResult)
+            })
+        })
+    })
+})
+```
+
+### Test Messages Are Specific ###
+
+Human-readable test messages must be clear, accurate, and specific. Some words
+essentially should never appear in these messages. For example:
+
+ - "properly"
+ - "correctly"
+ - "appropriately"
+
+The issue with these words in this context is that they obscure what the actual
+desired behaviour of the system is (it's obvious that a given function call
+must execute correctly -- what does this actually mean?).
```

Comment:
Is this saying we should be defining correctness in test descriptions?

I just worry that this could get overly-cumbersome. For example in this test: https://github.com/tracer-protocol/perpetual-pools-contracts/blob/pools-v2/test/SMAOracle.spec.ts#L132 to replace the use of vaguely referring to correctness, would you be expected to define how we implemented simple moving averages?

Reply:
> Is this saying we should be defining correctness in test descriptions?

It's saying that tests (and by extension, their associated messages) should be as specific as possible.

I wrote the test you linked and it's in violation of this guide. I would argue the word "correct" is entirely redundant in this instance. Deleting it would make this test compliant with this guide, in my opinion.

Reply:
With this test, for example: https://github.com/tracer-protocol/perpetual-pools-contracts/blob/pools-v2/test/PoolCommitter/commit.spec.ts#L1656

Should it be changed to `it("executes and updates balance", async () => {`, or should it define what correct execution and balance update entails?

I'm leaning towards the latter, but the changes in this PR include "it's obvious that a given function call
must execute correctly --  what does this actually mean?" which may be interpreted as I'm supposed to explain what correctness means for each test case that follows a happy path.